File: tlg/src/main/java/ch/bailu/tlg/State.java
package ch.bailu.tlg;

public abstract class State {
	protected final InternalContext context;
	
	
	
	public State(InternalContext c) {
		context=c;
	}
	
	
	public State init(PlatformContext c) {
		return this;
	}

	public State moveLeft(PlatformContext c) {
		return this;
	}

	
	public State moveRight(PlatformContext c) {
		return this;
	}
	
	public State moveDown(PlatformContext c) {
		return this;
	}

	public State moveTurn(PlatformContext c) {
		return this;
	}

	
	public State startNewGame(PlatformContext c) {
		return new StateRunning(context).init(c);
	}

	
	public State togglePause(PlatformContext c) {
		return this;
	}


	public static State restoreContextFactory(InternalContext c, int id) {
		if (id == StateRunning.ID) {
			return new StateRunning(c);
			
		} else if (id == StatePaused.ID) {
			return new StatePaused(c);
			
		} else if (id == StateHighscore.ID) {
			return new StateHighscore(c);
			
		} else {
			return new StateLocked(c);
		}
	}

	
	public static State gameOverStateFactory(InternalContext i, PlatformContext p) {
		HighscoreList highscoreList = new HighscoreList(p);
		
		if (highscoreList.haveNewHighscore(i.currentScore.getScore())) {
			return new StateHighscore(i).init(p);
			
		} else {
			return new StateLocked(i).init(p);
		}
		
	}
	

	public int getTimerInterval() {
		return 500;
	}


	
	public abstract int getID();


	public State setHighscoreName(PlatformContext c, String name) {
		return this;
	}

}

File: tlg/src/main/kotlin/state/State.kt
package state

import context.InternalContext
import context.PlatformContext
import score.HighscoreList

abstract class State(val context: InternalContext) {
    open fun init(c: PlatformContext): State {
        return this
    }

    open fun moveLeft(c: PlatformContext): State {
        return this
    }

    open fun moveRight(c: PlatformContext): State {
        return this
    }

    open fun moveDown(c: PlatformContext): State {
        return this
    }

    open fun moveTurn(c: PlatformContext): State {
        return this
    }

    fun startNewGame(c: PlatformContext): State {
        return StateRunning(context).init(c)
    }

    open fun togglePause(c: PlatformContext): State {
        return this
    }

    open val timerInterval: Int
        get() = 500

    abstract val id: Int

    open fun setHighscoreName(c: PlatformContext, name: String): State {
        return this
    }

    companion object {
        fun restoreContextFactory(iContext: InternalContext, id: Int): State {
            return if (id == StateRunning.ID) {
                StateRunning(iContext)
            } else if (id == StatePaused.ID) {
                StatePaused(iContext)
            } else if (id == StateHighscore.ID) {
                StateHighscore(iContext)
            } else {
                StateLocked(iContext)
            }
        }

        fun gameOverStateFactory(iContext: InternalContext, pContext: PlatformContext): State {
            val highscoreList = HighscoreList(pContext)
            return if (highscoreList.haveNewHighscore(iContext.currentScore.score)) {
                StateHighscore(iContext).init(pContext)
            } else {
                StateLocked(iContext).init(pContext)
            }
        }
    }
}
